File: app/src/main/java/com/proxerme/app/fragment/MainFragment.java
package com.proxerme.app.fragment;

import android.support.v4.app.Fragment;

import com.proxerme.app.activity.MainActivity;
import com.proxerme.app.application.MainApplication;
import com.proxerme.app.interfaces.OnActivityListener;
import com.proxerme.app.util.EventBusBuffer;

import org.greenrobot.eventbus.EventBus;

/**
 * TODO: Describe class
 *
 * @author Ruben Gees
 */
public abstract class MainFragment extends Fragment implements OnActivityListener {

    protected MainActivity getParentActivity() throws RuntimeException {
        try {
            return (MainActivity) getActivity();
        } catch (ClassCastException e) {
            throw new RuntimeException("Don't use this Fragment in another" +
                    " Activity than MainActivity.");
        }
    }

    protected final MainApplication getMainApplication() {
        return (MainApplication) getActivity().getApplication();
    }

    @Override
    public void onResume() {
        super.onResume();

        EventBus.getDefault().register(this);
        getEventBusBuffer().stopAndProcess();
    }

    @Override
    public void onPause() {
        getEventBusBuffer().startBuffering();
        EventBus.getDefault().unregister(this);

        super.onPause();
    }

    @Override
    public void onDestroy() {
        getEventBusBuffer().stopAndPurge();
        getMainApplication().getRefWatcher().watch(this);

        super.onDestroy();
    }

    @Override
    public boolean onBackPressed() {
        return false;
    }

    protected abstract EventBusBuffer getEventBusBuffer();

}

File: app/src/main/kotlin/com/proxerme/app/fragment/framework/MainFragment.kt
package com.proxerme.app.fragment.framework

import android.os.Bundle
import android.support.v4.app.Fragment
import com.proxerme.app.application.MainApplication
import com.proxerme.app.manager.SectionManager
import com.proxerme.app.util.extension.androidUri
import com.proxerme.app.util.extension.openHttpPage
import me.zhanghai.android.customtabshelper.CustomTabsHelperFragment
import okhttp3.HttpUrl

/**
 * TODO: Describe Class
 *
 * @author Ruben Gees
 */
abstract class MainFragment : Fragment() {

    abstract val section: SectionManager.Section

    private lateinit var customTabsHelper: CustomTabsHelperFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        customTabsHelper = CustomTabsHelperFragment.attachTo(this)
    }

    override fun onResume() {
        super.onResume()

        SectionManager.notifySectionResumed(section)
    }

    override fun onPause() {
        super.onPause()

        SectionManager.notifySectionPaused(section)
    }

    override fun onDestroy() {
        super.onDestroy()

        MainApplication.refWatcher.watch(this)
    }

    fun setLikelyUrl(url: HttpUrl) {
        customTabsHelper.mayLaunchUrl(url.androidUri(), Bundle(), emptyList())
    }

    fun showPage(url: HttpUrl) {
        customTabsHelper.openHttpPage(activity, url)
    }
}
